### DISCUSSION TITLE
Well-Known Tool Configurations

### DISCUSSION THREAD

**Discussion by halostatue** [twpayne/chezmoi#3280] [Ideas]
I wonder if it would be worthwhile adding support for well-known tool configurations. The default configuration for `merge` is `vimdiff` and the default configuration for `diff` is `diff -u`, but I wonder if it would be worthwhile to make it *easier* to support specific tools like Emacs, VS Code, Meld, Kaleidoscope, Beyond Compare, etc. Maybe instead of `merge.command` it would be `merge.tool` with a string-enum value that holds those. *Or* we add a new section to the documentation with configurations for those.

This is inspired by #3279, but is not *wholly* restricted to just `diff` and `merge`, although those are the only ones that I can think of because I just replied to that one.

The biggest argument *against* this would be deciding when a new tool would justify being added to either the `.tool` list or documentation. We don't necessarily want to support All The Tools, just the ones that are more likely to be used than not.

**Comment by twpayne** [comment]
Hmm, I think it would be sufficient to provide examples in the documentation of how set `merge.command` and `merge.args` for each tool. I don't think there needs to be any code change to chezmoi to support the different tools.

At a higher level, I'm keen that chezmoi provides mechanism, not policy. For example, chezmoi provides a mechanism to plug in external diff and merge tools, but (other than having sensible defaults) no policy limiting your choice of tool.

**Reply by halostatue** [reply]
Yes, it would be sufficient to document, but offering a `merge.tool` configuration would not necessarily be dictating *policy* so much as providing multiple named sensible defaults.

Nothing says that I couldn’t do `merge.tool = "emacs"` and `merge.command = "emacsclient"`. (It would make much less sense to provide a `merge.args` override in this case.)

I think that part of what I’m suggesting is an opportunity to make certain well-known configurations like this extensible.

Also, as an alternative, it *might* be nice to be able to do something like git does with `mergetool`:

```yaml
merge:
  tools:
    ksdiff:
      command: ksdiff
    emacs:
      command: emacs
      args: …
```

Then we could do `chezmoi merge --tool=emacs` or something similar so that I could support both command-line and GUI diffing or merging. This would be similar, I suspect with other external helper tools (pager, etc.).

**Reply by Lockszmith-GH** [reply]
(Hope it's ok to butt in):
How about letting git's configuration be the default behavior for `chezmoi`?

**Reply by halostatue** [reply]
I would probably be against this because git simply has `cmd` with environment variables as its parameters:

```ini
[difftool "Kaleidoscope"]
  cmd = "/usr/local/bin/ksdiff" --partial-changeset --relative-path "${MERGED}" -- "${LOCAL}" "${REMOTE}"

[mergetool "Kaleidoscope"]
  cmd = "/usr/local/bin/ksdiff" --merge --output "${MERGED}" --base "${BASE}" -- "${LOCAL}" --snapshot "${REMOTE}" --snapshot
  trustExitCode = true
```

Transforming that would be something of a pain, especially as chezmoi's triple for 3-way merge is `(Target,Source,Destination)` instead of `(Remote,Local,Base)`. The `Destination` isn't *quite* the same as `Base` or `Merged`.
